{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5573854f",
   "metadata": {},
   "source": [
    "## <center>Comparative validation of MAPLE and Viridian</center>\n",
    "##### <center>MANUAL</center>\n",
    "\n",
    "\n",
    "\n",
    "| **Label** | **start time** | **finish time** | **last modified** |\n",
    "|:--------------:|:-----------:|:-----------:|:----------------:|\n",
    "|   Project 1   |  2023-05-16 |  2023-07-26 |   2023-09-12     |\n",
    "\n",
    "\n",
    "&nbsp;\n",
    "\n",
    "```python\n",
    "# CODE: \n",
    "/nfs/research/goldman/zihao/Code/compProject_2_P2\n",
    "# DATA:\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2\n",
    "# Dir/to/plot/figure\n",
    "/nfs/research/goldman/zihao/Code/jupyterLab/Project_2_vircolComp/1.pievennPlot_P2\n",
    "```\n",
    "**Tipsï¼š** All the codes need to change the path has been marked in the code as below:\n",
    "  \n",
    "  \\# ==================Requires modification==================  \n",
    "  \\# Path to be changed  \n",
    "  \\# ==================Requires modification=================="
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9f98739",
   "metadata": {},
   "source": [
    "# <center>Catalog:</center>\n",
    "- 1. Data Preparation\n",
    "    - 1.1 [Extract assembled sequences](#1.1_Extract_assembled_sequences)\n",
    "    - 1.2 [Split processing](#1.2_Split_processing)\n",
    "    - 1.3 [Sequence alignment](#1.3_Sequence_alignment)\n",
    "\n",
    "&nbsp;\n",
    "\n",
    "- 2. [Data Processing](#2.Adding-info_gene-sequences)\n",
    "        \n",
    "&nbsp;\n",
    "\n",
    "- 3. [Calculating percentages](#3_calculate_infomations)\n",
    "    \n",
    "&nbsp;\n",
    "\n",
    "- 4. [Plotting](#4_visualization)\n",
    "    - 4.1 [For error positions](#4.1.For_err_pos)\n",
    "    - 4.2 [For all positions](#4.2.For_all_pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9846a69b",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Data Preparation\n",
    "## 1.1_Extract_assembled_sequences\n",
    "[Return to Catalog](#Catalog:)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b0e850c",
   "metadata": {},
   "source": [
    "- **Idea**: Using the IDs of the samples from Colman's assembly running MAPLE, extract the information of the sequences of the corresponding samples from Viridan assembly and save them.\n",
    "\n",
    "- **Input**: Mr Martin's assembled sequence (Viridian assemblies)\n",
    "```python\n",
    "/nfs/research/zi/mhunt/Viridian_wf_paper/Vdn_all_ena/Dataset.01/Batches.all_reads/$i.cons.fa.gz\n",
    "```\n",
    "\n",
    "- **Output**: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_1_dataPrep/folderData_extractSequence\n",
    "```\n",
    "\n",
    "- **Address**:\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_1extraceSequences/pg_for_extract.py\n",
    "```\n",
    "\n",
    "### Code block:\n",
    "###### Run the program\n",
    "```bash\n",
    "sh /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_1extraceSequences/bash_for_extract.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1b7cdc5",
   "metadata": {},
   "source": [
    "## 1.2_Split_processing\n",
    "[Return to Catalog](#Catalog:)\n",
    "- **Idea**: To facilitate sequence alignment, sequences are grouped together for subsequent parallelization\n",
    "\n",
    "- **Input**: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_1_dataPrep/folderData_extractSequence\n",
    "```\n",
    "- **Output**: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_1_dataPrep/folderData_extractSequence/split_files\n",
    "```\n",
    "- **Address**:\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_2splitProcessing/pg_splitProcessing.py\n",
    "```\n",
    "\n",
    "### Code block:\n",
    "###### Run the program\n",
    "```bash\n",
    "bsub -M 20000 -e /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_2splitProcessing/Folder_Checking/splitProcessing_errorChecking.txt 'python3 /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_2splitProcessing/pg_splitProcessing.py'\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4ac6dde",
   "metadata": {},
   "source": [
    "## 1.3_Sequence_alignment\n",
    "[Return to Catalog](#Catalog:)\n",
    "\n",
    "### 1. Sequence alignment using MAFFT software\n",
    "- **Run the program**:\n",
    "```bash\n",
    "bsub sh /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_3sequenceAlignment/Aligned.sh\n",
    "```\n",
    "\n",
    "### 2. split\n",
    "\n",
    "- **Address**: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_3sequenceAlignment/pg_splitAligned.py\n",
    "```\n",
    "- **Input**: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_1_dataPrep/Folder_alignment/output_${i}_aligned.fasta\n",
    "```\n",
    "- **Output**: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_1_dataPrep/Folder_alignment/folderData_alignedSplit\n",
    "```\n",
    "- **Run the program**:\n",
    "```bash\n",
    "sh /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_1dataPreparation/Folder_3sequenceAlignment/bash_splitAligned.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cfa3fb5",
   "metadata": {},
   "source": [
    "---\n",
    "# 2.Adding-info_gene-sequences\n",
    "[Return to Catalog](#Catalog:)\n",
    "- **Idea**:\n",
    "    - label_masked: The values in this column depend on whether the values in the nucleotide_martin and nucleotide_origin columns are within the specified character lists. If the value in the nucleotide_martin column is among ['-', 'n', 'm', 'r', 'w', 's', 'y', 'k', 'v', 'h', 'd', 'b'], and the value in nucleotide_origin is among ['-', 'n'], then the value in this column will be 1, otherwise 0.\n",
    "\n",
    "    - label_mar: The values in this column are determined by whether the value in the nucleotide_martin column is within the specified character list. If it's among ['-', 'n', 'm', 'r', 'w', 's', 'y', 'k', 'v', 'h', 'd', 'b'], then the value in this column will be 1, otherwise 0.\n",
    "\n",
    "    - label_ori: The values in this column depend on whether the value in the nucleotide_origin column is in the list ['-', 'n']. If it is, then the value in this column will be 1, otherwise 0.\n",
    "\n",
    "    - label_same: The values in this column depend on whether the values in the nucleotide_martin and nucleotide_origin columns are equal and whether they're within the specified character list. If they're equal and are in the list ['-', 'n', 'm', 'r', 'w', 's', 'y', 'k', 'v', 'h', 'd', 'b'], then the value in label_same will be 2. If they're equal but not in the specified list, then the value in label_same will be 1. If they're not equal, then it checks if the value in label_masked, label_mar, or label_ori is 1. If yes, then label_same will be 2, otherwise, it will be 0.\n",
    "    \n",
    "    - After that, add information about ANNOT and COV.\n",
    "\n",
    "- **Input**:   \n",
    "\n",
    "```python\n",
    "# Aligned sequences:\n",
    "## Viridian version: \n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_1_dataPrep/Folder_alignment/folderData_alignedSplit/\n",
    "## Colman version: \n",
    "/nfs/research/goldman/zihao/Datas/p1/File_5_consensus/Decompress/Aligned_split_July_for_MAPLE\n",
    "\n",
    "# Information about ANNOT and COV.\n",
    "## For Coverage:\n",
    "/nfs/research/goldman/zihao/Datas/p1/File_5_coverage/Decompress/{id}_coverage.txt\n",
    "## For Annotation:\n",
    "/nfs/research/goldman/zihao/Datas/p1/File_5_annot/Decompress/{id}_annot.txt\n",
    "```\n",
    "- **Output**: \n",
    "```python\n",
    "output path: /nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_2_combination/folderData_addInfo\n",
    "```\n",
    "- **Address**:\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_2addInfo/pg_mergeAllpos.py\n",
    "```\n",
    "\n",
    "### Code block:\n",
    "###### Run the program\n",
    "```bash\n",
    "sh /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_2addInfo/bash_combination.sh\n",
    "```\n",
    "#### Counting missing IDs for all postions: **(After finishing the above run)**\n",
    "```bash\n",
    "sh /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_2addInfo/bash_statMissingID.sh\n",
    "```\n",
    "================ Stat for combination ================  \n",
    "The number of missing IDs on merge is: 0  \n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_2_combination/missingFile_ids.txt  \n",
    "================ Stat for annot ================  \n",
    "The number of missing IDs for Annot is: 20  \n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_2_combination/missing_ann_ids.txt  \n",
    "================ Stat for cov ================  \n",
    "The number of missing IDs for COV is: 1444  \n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_2_combination/missing_cov_ids.txt  \n",
    "================ Stat total ================  \n",
    "The total number of missing IDs is: 1464"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e292f05",
   "metadata": {},
   "source": [
    "# 3_calculate_infomations\n",
    "[Return to Catalog](#Catalog:)\n",
    "### Code block:\n",
    "###### Run the program\n",
    "```bash\n",
    "sh /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/bash_calcul.sh\n",
    "```\n",
    "### Or enter the commends step by stepï¼š\n",
    "### Part1_for_err_pos\n",
    "- **Input**: \n",
    "```python\n",
    "# MAPLE's output: (a_file_path)\n",
    "'/nfs/research/goldman/zihao/Datas/p2_compViridian_P3/Folder_mapleOutput/COL_output_modified.txt'\n",
    "# Files to be processed: (b_folder_path)\n",
    "\"/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_2_combination/folderData_addInfo\"\n",
    "```\n",
    "- **Output**: \n",
    "```python\n",
    "output_file_path = '/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/errPos/errpos_outputData.txt'\n",
    "missing_b_file_path = '/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/errPos/errPos_missingFiles.txt'\n",
    "missing_rows_file_path = '/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/errPos/errPos_missingPos.txt'\n",
    "\n",
    "```\n",
    "- **Address**:\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/errPos.py\n",
    "```\n",
    "\n",
    "### Code block:\n",
    "###### Run the program\n",
    "```bash\n",
    "bsub -M 2000 -e /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/Folder_Checking/errPos_errorChecking.txt -o /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/Folder_Checking/errPos_outputChecking.txt 'python3 /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/errPos.py'\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16ed1605",
   "metadata": {},
   "source": [
    "### Part2_for_all_pos\n",
    "#### 1.For venn chart\n",
    "##### diff\n",
    "```bash\n",
    "bsub -M 5000 -e /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/Folder_Checking/vennDiff_allPos_errorChecking.txt -o /nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_vennDiff.txt 'python3 /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/allPOS_diff.py'\n",
    "```\n",
    "##### masked\n",
    "```bash\n",
    "bsub -M 5000 -e /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/Folder_Checking/vennMasked_allPos_errorChecking.txt -o /nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_vennMasked.txt 'python3 /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/allPOS_masked.py'\n",
    "```\n",
    "##### same\n",
    "```bash\n",
    "bsub -M 5000 -e /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/Folder_Checking/vennSame_allPos_errorChecking.txt -o /nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_vennSame.txt 'python3 /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/allPOS_same.py'\n",
    "```\n",
    "#### 2.For pie chart\n",
    "```bash\n",
    "bsub -M 5000 -e /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/Folder_Checking/pieAllpos_errorChecking.txt -o /nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_pie.txt 'python3 /nfs/research/goldman/zihao/Code/compProject_2_P2/Folder_3calculing/allPOS_pie.py'\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5156bda6",
   "metadata": {},
   "source": [
    "# 4_visualization\n",
    "[Return to Catalog](#Catalog:)\n",
    "\n",
    "- Data preparation + For pie chart: ==========ã€‹\n",
    "### 4.1.For_err_pos\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/jupyterLab/Project_2_vircolComp/1.pievennPlot_P2/1.1.pieChart_ERR.ipynb\n",
    "```\n",
    "\n",
    "- For venn chart: ==========ã€‹\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/jupyterLab/Project_2_vircolComp/1.pievennPlot_P2/1.2.vennChart_ERR.Rmd\n",
    "```\n",
    "### 4.2.For_all_pos\n",
    "    - 1.Extract data from the following output\n",
    "```python\n",
    "# 1.1 forVenn_diff\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_vennDiff.txt\n",
    "# 1.2 forVenn_masked\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_vennMasked.txt\n",
    "# 1.3 forVenn_same\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_vennSame.txt\n",
    "# 2 forPie\n",
    "/nfs/research/goldman/zihao/Datas/p2_compViridian_P2/Folder_3_calculForPLOT/allPos/allPos_pie.txt\n",
    "```\n",
    "    - 2.Inputting data into ipy for processing\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/jupyterLab/Project_2_vircolComp/1.pievennPlot_P2/2.1.pieChart_ALL.ipynb\n",
    "```\n",
    "\n",
    "- For venn chart: ==========ã€‹\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/jupyterLab/Project_2_vircolComp/1.pievennPlot_P2/2.2.vennChart_ALL.Rmd\n",
    "```\n",
    "\n",
    "### Final - Combination:\n",
    "```python\n",
    "/nfs/research/goldman/zihao/Code/jupyterLab/Project_2_vircolComp/1.pievennPlot_P2/FINAL-plotVenn.vsdx\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
