{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c252019b",
   "metadata": {},
   "source": [
    "- 4. what Nick was suggesting, that is, using the 180,000 genomes from 1., look at the AF of the reference alleles and Colman consensus’ alleles. Of course the same analysis could be done with Viridian’s assemblies as well, but we can start with Colman’s.\n",
    "\n",
    "- This time you would need to look only at the position that have an alternative allele, so AF>1% or AF>0% or something like that.\n",
    "    - For each of these positions, you check which are the two nucleotides, let’s call them N1 and N2 at frequencies F1 and F2=100%-F1\n",
    "    - And then you need to check which of these two nucleotides is the reference nucleotide at the considered position, and which one is the nucleotide at Colman’s assembly at this position.\n",
    "- We want to see what frequencies have the reference nucleotides and what frequencies have Colman’s assembly nucleotides.\n",
    "- It might be that Colman’s assembly nucleotides are always >50%, but I don’t know for sure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4c4b8a5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>SB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.283465</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>71</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.028075</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>241</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001150</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>600</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001566</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29740</th>\n",
       "      <td>29741</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000978</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29747</th>\n",
       "      <td>29748</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001205</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29768</th>\n",
       "      <td>29769</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001678</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29803</th>\n",
       "      <td>29804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003504</td>\n",
       "      <td>656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29837</th>\n",
       "      <td>29838</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.452229</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         POS REF ALT        AF   SB\n",
       "5          6   A   G  0.466667    3\n",
       "20        21   C   T  0.283465    3\n",
       "70        71   C   T  0.028075   49\n",
       "240      241   C   T  0.001150    2\n",
       "599      600   G   A  0.001566    0\n",
       "...      ...  ..  ..       ...  ...\n",
       "29740  29741   C   T  0.000978    2\n",
       "29747  29748   T   C  0.001205    3\n",
       "29768  29769   C   A  0.001678    0\n",
       "29803  29804   T   C  0.003504  656\n",
       "29837  29838   C   T  0.452229   42\n",
       "\n",
       "[236 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('/nfs/research/goldman/zihao/Datas/p1/File_5_annot/New_Decompress/SRR20493490_annot.txt', \n",
    "                 sep='\\t')\n",
    "df = df[df['AF'] > 0]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1b93834",
   "metadata": {},
   "source": [
    "REF: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/errorsProject_1/Consensuses/ref_MN908947.3.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0369f5e5",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "fasta_file = \"/nfs/research/goldman/zihao/errorsProject_1/Consensuses/ref_MN908947.3.fasta\"\n",
    "\n",
    "def read_fasta_file(file_path):\n",
    "    sequences = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        current_sequence = \"\"\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if current_sequence:\n",
    "                    sequences.append(current_sequence)\n",
    "                    current_sequence = \"\"\n",
    "            else:\n",
    "                current_sequence += line\n",
    "        if current_sequence:\n",
    "            sequences.append(current_sequence)\n",
    "    return sequences\n",
    "\n",
    "sequences = read_fasta_file(fasta_file)\n",
    "\n",
    "data = {'Position': [], 'Nucleotide': []}\n",
    "\n",
    "for sequence in sequences:\n",
    "    nucleotides = list(sequence)\n",
    "    positions = list(range(1, len(sequence) + 1))\n",
    "    data['Position'].extend(positions)\n",
    "    data['Nucleotide'].extend(nucleotides)\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "\n",
    "# 保存为 CSV 文件\n",
    "df.to_csv('/nfs/research/goldman/zihao/errorsProject_1/Consensuses/ref_MN908947.3.txt', \n",
    "          sep='\\t', index=False)\n",
    "df\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "59bcfdb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Nucleotide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29898</th>\n",
       "      <td>29899</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29899</th>\n",
       "      <td>29900</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29900</th>\n",
       "      <td>29901</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29901</th>\n",
       "      <td>29902</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29902</th>\n",
       "      <td>29903</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29903 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position Nucleotide\n",
       "0             1          A\n",
       "1             2          T\n",
       "2             3          T\n",
       "3             4          A\n",
       "4             5          A\n",
       "...         ...        ...\n",
       "29898     29899          A\n",
       "29899     29900          A\n",
       "29900     29901          A\n",
       "29901     29902          A\n",
       "29902     29903          A\n",
       "\n",
       "[29903 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ref = pd.read_csv('/nfs/research/goldman/zihao/errorsProject_1/Consensuses/ref_MN908947.3.txt', \n",
    "          sep='\\t')\n",
    "df_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c254e10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['POS', 'REF', 'ALT', 'AF', 'SB', 'Position', 'Nucleotide'], dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "80fd671c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>Nucleotide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.283465</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>71</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.028075</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>241</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001150</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>600</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001566</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>29741</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000978</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>29748</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001205</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>29769</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001678</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>29804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003504</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>29838</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.452229</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       POS REF ALT        AF Nucleotide\n",
       "0        6   A   G  0.466667          A\n",
       "1       21   C   T  0.283465          C\n",
       "2       71   C   T  0.028075          C\n",
       "3      241   C   T  0.001150          C\n",
       "4      600   G   A  0.001566          G\n",
       "..     ...  ..  ..       ...        ...\n",
       "231  29741   C   T  0.000978          C\n",
       "232  29748   T   C  0.001205          T\n",
       "233  29769   C   A  0.001678          C\n",
       "234  29804   T   C  0.003504          T\n",
       "235  29838   C   T  0.452229          C\n",
       "\n",
       "[236 rows x 5 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(df, df_ref, left_on='POS', right_on='Position')[['POS', 'REF', 'ALT', 'AF', 'Nucleotide']]\n",
    "\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f256b6e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>Nucleotide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [POS, REF, ALT, AF, Nucleotide]\n",
       "Index: []"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[merged_df['REF'] != merged_df['Nucleotide']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c2a9961",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "587b7e21",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3779ed4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "634b7ca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "fasta_file = \"/nfs/research/goldman/zihao/errorsProject_1/Consensuses/ref_MN908947.3.fasta\"\n",
    "\n",
    "def read_fasta_file(file_path):\n",
    "    sequences = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        current_sequence = \"\"\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if current_sequence:\n",
    "                    sequences.append(current_sequence)\n",
    "                    current_sequence = \"\"\n",
    "            else:\n",
    "                current_sequence += line\n",
    "        if current_sequence:\n",
    "            sequences.append(current_sequence)\n",
    "    return sequences\n",
    "\n",
    "sequences = read_fasta_file(fasta_file)\n",
    "\n",
    "data = {'Position': [], 'Nucleotide': []}\n",
    "\n",
    "for sequence in sequences:\n",
    "    nucleotides = list(sequence)\n",
    "    positions = list(range(1, len(sequence) + 1))\n",
    "    data['Position'].extend(positions)\n",
    "    data['Nucleotide'].extend(nucleotides)\n",
    "\n",
    "df = pd.DataFrame(data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
