{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c252019b",
   "metadata": {},
   "source": [
    "- 4. what Nick was suggesting, that is, using the 180,000 genomes from 1., look at the AF of the reference alleles and Colman consensus’ alleles. Of course the same analysis could be done with Viridian’s assemblies as well, but we can start with Colman’s.\n",
    "\n",
    "- This time you would need to look only at the position that have an alternative allele, so AF>1% or AF>0% or something like that.\n",
    "    - For each of these positions, you check which are the two nucleotides, let’s call them N1 and N2 at frequencies F1 and F2=100%-F1\n",
    "    - And then you need to check which of these two nucleotides is the reference nucleotide at the considered position, and which one is the nucleotide at Colman’s assembly at this position.\n",
    "- We want to see what frequencies have the reference nucleotides and what frequencies have Colman’s assembly nucleotides.\n",
    "- It might be that Colman’s assembly nucleotides are always >50%, but I don’t know for sure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6fa1ed",
   "metadata": {},
   "source": [
    "New_Folder: \n",
    "```python\n",
    "/nfs/research/goldman/zihao/Datas/p1/File_5_annot/Folder_Decompress_forP4/\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5c6d57bc",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>AF2</th>\n",
       "      <th>nucleotide_martin</th>\n",
       "      <th>nucleotide_origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>373</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.994845</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>378</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003417</td>\n",
       "      <td>0.996583</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>443</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.003759</td>\n",
       "      <td>0.996241</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>491</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.004795</td>\n",
       "      <td>0.995205</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>514</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.048805</td>\n",
       "      <td>0.951195</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>28541</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>0.997909</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>28602</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.003405</td>\n",
       "      <td>0.996595</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>28703</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.002670</td>\n",
       "      <td>0.997330</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>29563</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.006442</td>\n",
       "      <td>0.993558</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>29625</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.006029</td>\n",
       "      <td>0.993971</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>226 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       POS REF ALT        AF       AF2 nucleotide_martin nucleotide_origin\n",
       "0      373   G   T  0.005155  0.994845                 G                 G\n",
       "1      378   T   C  0.003417  0.996583                 T                 T\n",
       "2      443   G   A  0.003759  0.996241                 G                 G\n",
       "3      491   G   T  0.004795  0.995205                 G                 G\n",
       "4      514   T   G  0.048805  0.951195                 T                 T\n",
       "..     ...  ..  ..       ...       ...               ...               ...\n",
       "221  28541   G   T  0.002091  0.997909                 G                 G\n",
       "222  28602   T   A  0.003405  0.996595                 T                 T\n",
       "223  28703   G   T  0.002670  0.997330                 G                 G\n",
       "224  29563   C   A  0.006442  0.993558                 C                 C\n",
       "225  29625   C   T  0.006029  0.993971                 C                 C\n",
       "\n",
       "[226 rows x 7 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 读取第一个文件\n",
    "df1 = pd.read_csv('/nfs/research/goldman/zihao/Datas/p1/File_5_annot/New_Decompress/ERR6774728_annot.txt', delimiter='\\t')\n",
    "# 读取第二个文件\n",
    "df2 = pd.read_csv('/nfs/research/goldman/zihao/Datas/p2_comp_viridian/3_combination/ERR6774728.txt', delimiter='\\t')\n",
    "\n",
    "df1 = df1[df1['AF'] > 0].drop('SB', axis=1)\n",
    "\n",
    "# 应用函数并添加ALT_2列\n",
    "# df1['ALT2'] = df1['ALT'].apply(set_ALT_2)\n",
    "df1['AF2'] = 1 - df1['AF']\n",
    "# 使用POS和position列合并两个数据框，用\"inner\"参数只保留两个表都有的行\n",
    "df = pd.merge(df1, df2, left_on='POS', right_on='position', how='inner')\n",
    "\n",
    "# 仅保留你需要的列\n",
    "df = df[['POS', 'REF', 'ALT', 'AF', 'AF2', 'nucleotide_martin', 'nucleotide_origin']]\n",
    "df[['nucleotide_martin', 'nucleotide_origin']] = df[['nucleotide_martin', 'nucleotide_origin']].applymap(lambda x: x.upper() if isinstance(x, str) else x)\n",
    "\n",
    "# 将结果写入新的txt文件，使用tab分隔符，不要写入索引\n",
    "#df.to_csv('/nfs/research/goldman/zihao/Datas/merged.txt', sep='\\t', index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b82b703",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "058786f9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f18cbbca",
   "metadata": {},
   "source": [
    "# TODO:\n",
    "- check_id select"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4710e087",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "# 获取文件夹1和文件夹2中的所有文件路径\n",
    "folder1_files = glob.glob('/nfs/research/goldman/zihao/Datas/p1/File_5_annot/New_Decompress/*.txt')\n",
    "folder2_files = glob.glob('/nfs/research/goldman/zihao/Datas/p2_comp_viridian/3_combination/*.txt')\n",
    "\n",
    "# 为文件夹2中的文件创建一个映射，以便我们可以通过ID查找它们\n",
    "folder2_files_dict = {os.path.basename(path).split('.')[0]: path for path in folder2_files}\n",
    "\n",
    "# 遍历文件夹1中的所有文件\n",
    "for file1 in folder1_files:\n",
    "    file1_id = os.path.basename(file1).split('_')[0]\n",
    "\n",
    "    # 检查我们是否在文件夹2中找到了相同ID的文件\n",
    "    if file1_id not in folder2_files_dict:\n",
    "        print(f\"No matching file in folder 2 for ID: {file1_id}\")\n",
    "        continue\n",
    "\n",
    "    file2 = folder2_files_dict[file1_id]\n",
    "\n",
    "    # 读取文件\n",
    "    df1 = pd.read_csv(file1, delimiter='\\t')\n",
    "    df2 = pd.read_csv(file2, delimiter='\\t')\n",
    "\n",
    "    # 你的数据处理步骤在这里\n",
    "    df1 = df1[df1['AF'] > 0].drop('SB', axis=1)\n",
    "    df1['AF2'] = 1 - df1['AF']\n",
    "    df = pd.merge(df1, df2, left_on='POS', right_on='position', how='inner')\n",
    "    df = df[['POS', 'REF', 'ALT', 'AF', 'AF2', 'nucleotide_martin', 'nucleotide_origin']]\n",
    "    df[['nucleotide_martin', 'nucleotide_origin']] = df[['nucleotide_martin', 'nucleotide_origin']].applymap(lambda x: x.upper() if isinstance(x, str) else x)\n",
    "\n",
    "    # 将结果写入新的txt文件，使用tab分隔符，不要写入索引\n",
    "    df.to_csv(f'/nfs/research/goldman/zihao/Datas/merged_{file1_id}.txt', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41eae8a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
